File: interview150/go/array/mergeSortedArray.go
package main

import "fmt"

func merge(nums1 []int, m int, nums2 []int, n int) {
	for n != 0 {
		if m != 0 && nums1[m-1] > nums2[n-1] {

			nums1[m+n-1] = nums1[m-1]
			m--
		} else {
			nums1[m+n-1] = nums2[n-1]
			n--
		}
	}
}

func main() {
	nums1 := []int{1, 2, 3, 0, 0, 0}
	m := 3
	nums2 := []int{2, 5, 6}
	n := 3
	merge(nums1, m, nums2, n)

	fmt.Println(nums1)
	nums1 = []int{1}
	m = 1
	nums2 = []int{}
	n = 0
	merge(nums1, m, nums2, n)
	fmt.Println(nums1)

	nums1 = []int{7, 8, 9, 0, 0, 0}
	m = 3
	nums2 = []int{1, 2, 3}
	n = 3
	merge(nums1, m, nums2, n)
	fmt.Println(nums1)

}

File: interview150Helped/py/mergeSortedArray.py
from typing import List
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        i,j=0,0
        while i<m or j<n:
            if j<n and i<m:
                if nums2[j] < nums1[i]:
                    nums1[i+1:m-1] = nums1[i:m]
                    nums1[i] = nums2[j]

                    j+=1
                    m+=1
                else:
                    i+=1
            else:
                if j<n:
                    nums1[i] = nums2[j]
                    j+=1
                    i+=1
                else:
                    break


        print(nums1)




if __name__ == "__main__":
    nums1 = [1,2,3,0,0,0]
    m=3
    nums2 = [2,5,6]
    n=3
    s = Solution()
    s.merge(nums1,m,nums2,n)